### PR TITLE
Disable spill for source stage broadcast join

### PR DESCRIPTION
**Pull request by sopel39**
Inserting local exchanges for such joins can dramatically hurt performance when spill is not triggered.

### CODE REVIEW

**Comment by sopel39** [comment]
Performance of spill (but not triggered) vs no spill is indifferent (this is desirable):
obraz
obraz
obraz
obraz
obraz
obraz
obraz

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalExecutionPlanner.java [hunk 1]
```diff
@@ -1584,7 +1584,7 @@ public PhysicalOperation visitIndexJoin(IndexJoinNode node, LocalExecutionPlanCo
             }
 
             OperatorFactory lookupJoinOperatorFactory;
-            OptionalInt totalOperatorsCount = getJoinOperatorsCountForSpill(context, session);
+            OptionalInt totalOperatorsCount = context.getDriverInstanceCount();
```
Comment (dain) [inline presto-main/src/main/java/io/prestosql/sql/planner/LocalExecutionPlanner.java]: Typo in commit message "Do not spill **fort** join over table scan"

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalExecutionPlanner.java [hunk 2]
```diff
@@ -1584,7 +1584,7 @@ public PhysicalOperation visitIndexJoin(IndexJoinNode node, LocalExecutionPlanCo
             }
 
             OperatorFactory lookupJoinOperatorFactory;
-            OptionalInt totalOperatorsCount = getJoinOperatorsCountForSpill(context, session);
```
Comment (dain) [inline presto-main/src/main/java/io/prestosql/sql/planner/LocalExecutionPlanner.java]: Was this a mistake in the existing code?  I don't think index join needs or supports spilling anyway.

Comment (findepi) [reply presto-main/src/main/java/io/prestosql/sql/planner/LocalExecutionPlanner.java]: Good catch. @sopel39 please extract this change to a separate commit.

**Comment by sopel39** [comment]
ac, @findepi could you take a look?

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalExecutionPlanner.java [hunk 3]
```diff
@@ -2100,7 +2102,10 @@ private OperatorFactory createLookupJoin(
             List<Integer> probeJoinChannels = ImmutableList.copyOf(getChannelsForSymbols(probeSymbols, probeSource.getLayout()));
             OptionalInt probeHashChannel = probeHashSymbol.map(channelGetter(probeSource))
                     .map(OptionalInt::of).orElse(OptionalInt.empty());
-            OptionalInt totalOperatorsCount = getJoinOperatorsCountForSpill(context, session);
+            OptionalInt totalOperatorsCount = context.getDriverInstanceCount();
+            if (spillEnabled) {
+                checkState(totalOperatorsCount.isPresent(), "A fixed distribution is required for JOIN when spilling is enabled");
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/sql/planner/LocalExecutionPlanner.java]: `checkState( !spillEnabled || totalOperatorsCount.isPresent(), ...)` (ie without `if`)

File: presto-main/src/main/java/io/prestosql/sql/planner/optimizations/AddLocalExchanges.java [hunk 4]
```diff
@@ -576,18 +577,21 @@ public PlanWithProperties visitUnion(UnionNode node, StreamPreferredProperties p
...
-                        fixedParallelism(),
-                        parentPreferences.constrainTo(node.getLeft().getOutputSymbols()).withFixedParallelism());
-            }
-            else {
-                probe = planAndEnforce(
-                        node.getLeft(),
-                        defaultParallelism(session),
-                        parentPreferences.constrainTo(node.getLeft().getOutputSymbols()).withDefaultParallelism(session));
+                if (probe.getProperties().getDistribution() != FIXED) {
+                    // Disable spill for joins over non-fixed streams as otherwise we would need to inert local exchange.
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/sql/planner/optimizations/AddLocalExchanges.java]: inert -> insert

File: presto-main/src/main/java/io/prestosql/sql/planner/optimizations/AddLocalExchanges.java [hunk 5]
```diff
@@ -576,18 +577,21 @@ public PlanWithProperties visitUnion(UnionNode node, StreamPreferredProperties p
...
+                    // Such local exchanges can dramatically hurt performance when spill is not triggered.
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/sql/planner/optimizations/AddLocalExchanges.java]: remove "dramatically". Your explanation is sufficient without it

File: presto-main/src/main/java/io/prestosql/sql/planner/optimizations/AddLocalExchanges.java [hunk 6]
```diff
@@ -576,18 +577,21 @@ public PlanWithProperties visitUnion(UnionNode node, StreamPreferredProperties p
...
+                    // When spill is not triggered it should not introduce performance penalty.
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/sql/planner/optimizations/AddLocalExchanges.java]: introduce -> induce

File: presto-main/src/main/java/io/prestosql/sql/planner/optimizations/AddLocalExchanges.java [hunk 7]
```diff
@@ -600,7 +604,6 @@ public PlanWithProperties visitJoin(JoinNode node, StreamPreferredProperties par
                 buildPreference = singleStream();
             }
             PlanWithProperties build = planAndEnforce(node.getRight(), buildPreference, buildPreference);
-
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/sql/planner/optimizations/AddLocalExchanges.java]: revert please

File: presto-main/src/test/java/io/prestosql/sql/planner/optimizations/TestAddExchangesPlans.java [hunk 8]
```diff
@@ -16,23 +16,60 @@
...
+    }
+
+    private static LocalQueryRunner createQueryRunner()
+    {
+        Session session = testSessionBuilder()
+                .setCatalog("tpch")
+                .setSchema("tiny")
+                .build();
+        FeaturesConfig featuresConfig = new FeaturesConfig()
+                .setSpillerSpillPaths("/tmp/spill_path");
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/sql/planner/optimizations/TestAddExchangesPlans.java]: Will this path be created?
If not, use something like "/tmp/some/nonexistent/path"
if yes, you need this to be random, and do cleanup.

File: presto-main/src/test/java/io/prestosql/sql/planner/optimizations/TestAddExchangesPlans.java [hunk 9]
```diff
@@ -95,4 +132,42 @@ public void testRepartitionForUnionAllBeforeHashJoin()
...
+                                                anyTree(
+                                                        tableScan("region", ImmutableMap.of("regionkey", "regionkey"))))))));
+    }
+
+    private Session spillEnabled(JoinDistributionType joinDistributionType)
+    {
+        return Session.builder(getQueryRunner().getDefaultSession())
+                .setSystemProperty(JOIN_DISTRIBUTION_TYPE, joinDistributionType.toString())
+                .setSystemProperty(SPILL_ENABLED, "true")
+                .setSystemProperty(TASK_CONCURRENCY, "16")
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/sql/planner/optimizations/TestAddExchangesPlans.java]: does it matter?

File: presto-main/src/test/java/io/prestosql/sql/planner/optimizations/TestAddExchangesPlans.java [hunk 10]
```diff
@@ -95,4 +132,42 @@ public void testRepartitionForUnionAllBeforeHashJoin()
                                                 anyTree(
                                                         tableScan("region", ImmutableMap.of("regionkey", "regionkey"))))))));
     }
+
+    @Test
+    public void testNonSpillableBroadcastJoinAboveTableScan()
+    {
+        assertDistributedPlan(
+                "SELECT * FROM nation n join region r on n.nationkey = r.regionkey",
+                spillEnabled(BROADCAST),
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/sql/planner/optimizations/TestAddExchangesPlans.java]: It looks like you're enabling spill for broadcast joins.
inline (or rename, if you can find a good name)

File: presto-main/src/main/java/io/prestosql/sql/planner/LocalExecutionPlanner.java (hunk 1 above)
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/sql/planner/LocalExecutionPlanner.java]: since Lookup Join does not support spill, would be good not to pass `partitioningSpillerFactory` as well.
Eg.

```
// There is no spill in lookup join's case
OptionalInt totalOperatorsCount = context.getDriverInstanceCount();
PartitioningSpillerFactory partitioningSpillerFactory = unsupportedPartitioningSpillerFactory();
```

Comment (findepi) [reply presto-main/src/main/java/io/prestosql/sql/planner/LocalExecutionPlanner.java]: Also, i am not sure about cmt msg "Fix index join obtaining driver instance count when spill is enabled". Can you rephrase it? From what i can understand, you're fixing general bug, as query with index join right above table scan and with spill enabled could fail. If that is correct then this would be more appropriate:
```
Fix query failure when spill enabled and there is Index Join above Table Scan
```

**Comment by findepi** [comment]
@sopel39 perhaps https://prestosql.io/docs/current/admin/spill.html#supported-operations needs to be updated
